File: dart/lib/src/json_read.dart
/// Field readers shared by the seed, preset and options parsers.
///
/// A document from the network is untrusted, so no reader here ever coerces.
/// `"minSeverity": "3"` is a document the sender got wrong, and saying so is
/// more useful than guessing what was meant.
///
/// Two rules hold for every reader:
///
///  * A missing field and an explicit `null` mean the same thing. Both leave
///    the caller's default in place.
///  * A field that is present with the wrong type throws [FormatException],
///    naming the field and the type the schema wants.
///
/// `dotnet/src/Vulgarity/JsonRead.cs` is the mirror of this file.
library;

/// Names the JSON type of [value] for an error message.
String jsonTypeName(Object? value) {
  if (value == null) {
    return 'null';
  }
  if (value is String) {
    return 'a string';
  }
  if (value is bool) {
    return 'a boolean';
  }
  if (value is num) {
    return 'a number';
  }
  if (value is List) {
    return 'an array';
  }
  if (value is Map) {
    return 'an object';
  }
  return 'a $value';
}

/// Reads a whole number. Returns null when the field is absent.
///
/// A fractional number is refused, so `1.0` is not a stand-in for `1`. On
/// dart2js every number is a double and `1.0` is indistinguishable from `1`,
/// which is a limit of the runtime rather than of this check.
int? readOptionalInt(Map<String, dynamic> json, String field) {
  final Object? value = json[field];
  if (value == null) {
    return null;
  }
  if (value is! int) {
    throw FormatException("'$field' must be a whole number. "
        'The document states ${jsonTypeName(value)}.');
  }
  return value;
}

/// Reads a boolean. Returns null when the field is absent.
bool? readOptionalBool(Map<String, dynamic> json, String field) {
  final Object? value = json[field];
  if (value == null) {
    return null;
  }
  if (value is! bool) {
    throw FormatException("'$field' must be true or false. "
        'The document states ${jsonTypeName(value)}.');
  }
  return value;
}

/// Reads a string. Returns null when the field is absent.
String? readOptionalString(Map<String, dynamic> json, String field) {
  final Object? value = json[field];
  if (value == null) {
    return null;
  }
  if (value is! String) {
    throw FormatException("'$field' must be a string. "
        'The document states ${jsonTypeName(value)}.');
  }
  return value;
}

/// Reads an array of strings. A missing field returns an empty list.
///
/// An empty string carries no meaning in any of these lists, so it drops out
/// rather than failing. A value of another type is a malformed document.
List<String> readStrings(Map<String, dynamic> json, String field) {
  final Object? value = json[field];
  if (value == null) {
    return const <String>[];
  }
  if (value is! List) {
    throw FormatException("'$field' must be an array of strings. "
        'The document states ${jsonTypeName(value)}.');
  }

  final List<String> result = <String>[];
  for (final Object? item in value) {
    if (item is! String) {
      throw FormatException("'$field' must hold strings only. "
          'It holds ${jsonTypeName(item)}.');
    }
    if (item.isNotEmpty) {
      result.add(item);
    }
  }
  return result;
}

File: dotnet/src/Vulgarity/JsonRead.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Vulgarity
{
    /// <summary>Field readers shared by the seed, preset and options parsers.</summary>
    /// <remarks>
    /// <para>
    /// A document from the network is untrusted, so no reader here ever coerces.
    /// <c>"minSeverity": "3"</c> is a document the sender got wrong, and saying so
    /// is more useful than guessing what was meant.
    /// </para>
    /// <para>
    /// Two rules hold for every reader. A missing field and an explicit
    /// <c>null</c> mean the same thing: both leave the caller's default in place.
    /// A field that is present with the wrong type throws
    /// <see cref="FormatException"/>, naming the field and the type the schema
    /// wants.
    /// </para>
    /// <para>dart/lib/src/json_read.dart is the mirror of this file.</para>
    /// </remarks>
    internal static class JsonRead
    {
        /// <summary>Names the JSON type of a value for an error message.</summary>
        public static string TypeName(JsonElement value)
        {
            switch (value.ValueKind)
            {
                case JsonValueKind.Undefined:
                case JsonValueKind.Null:
                    return "null";
                case JsonValueKind.String:
                    return "a string";
                case JsonValueKind.True:
                case JsonValueKind.False:
                    return "a boolean";
                case JsonValueKind.Number:
                    return "a number";
                case JsonValueKind.Array:
                    return "an array";
                case JsonValueKind.Object:
                    return "an object";
                default:
                    return "a value";
            }
        }

        /// <summary>Reads a whole number. Returns null when the field is absent.</summary>
        /// <remarks>
        /// A fractional number is refused, so <c>1.0</c> is not a stand-in for
        /// <c>1</c>.
        /// </remarks>
        public static int? ReadOptionalInt(JsonElement json, string field)
        {
            JsonElement value;
            if (!json.TryGetProperty(field, out value) || value.ValueKind == JsonValueKind.Null)
            {
                return null;
            }

            int result;
            if (value.ValueKind != JsonValueKind.Number || !value.TryGetInt32(out result))
            {
                throw new FormatException(
                    "'" + field + "' must be a whole number. The document states " + TypeName(value) + ".");
            }

            return result;
        }

        /// <summary>Reads a boolean. Returns null when the field is absent.</summary>
        public static bool? ReadOptionalBool(JsonElement json, string field)
        {
            JsonElement value;
            if (!json.TryGetProperty(field, out value) || value.ValueKind == JsonValueKind.Null)
            {
                return null;
            }

            if (value.ValueKind != JsonValueKind.True && value.ValueKind != JsonValueKind.False)
            {
                throw new FormatException(
                    "'" + field + "' must be true or false. The document states " + TypeName(value) + ".");
            }

            return value.ValueKind == JsonValueKind.True;
        }

        /// <summary>Reads a string. Returns null when the field is absent.</summary>
        public static string ReadOptionalString(JsonElement json, string field)
        {
            JsonElement value;
            if (!json.TryGetProperty(field, out value) || value.ValueKind == JsonValueKind.Null)
            {
                return null;
            }

            if (value.ValueKind != JsonValueKind.String)
            {
                throw new FormatException(
                    "'" + field + "' must be a string. The document states " + TypeName(value) + ".");
            }

            return value.GetString();
        }

        /// <summary>Reads an array of strings. A missing field returns an empty list.</summary>
        /// <remarks>
        /// An empty string carries no meaning in any of these lists, so it drops out
        /// rather than failing. A value of another type is a malformed document.
        /// </remarks>
        public static List<string> ReadStrings(JsonElement json, string field)
        {
            List<string> result = new List<string>();

            JsonElement value;
            if (!json.TryGetProperty(field, out value) || value.ValueKind == JsonValueKind.Null)
            {
                return result;
            }

            if (value.ValueKind != JsonValueKind.Array)
            {
                throw new FormatException(
                    "'" + field + "' must be an array of strings. The document states " + TypeName(value) + ".");
            }

            foreach (JsonElement item in value.EnumerateArray())
            {
                if (item.ValueKind != JsonValueKind.String)
                {
                    throw new FormatException(
                        "'" + field + "' must hold strings only. It holds " + TypeName(item) + ".");
                }

                string text = item.GetString();
                if (!string.IsNullOrEmpty(text))
                {
                    result.Add(text);
                }
            }

            return result;
        }
    }
}
